File: auth_service/src/dbq.ts
import { ResultSetHeader, RowDataPacket } from "mysql2/promise";
import { v4 as uuidv4 } from "uuid";

import { AccountResponse } from "./types";

interface DbExecutor {
  execute<T extends ResultSetHeader | RowDataPacket[]>(
    sql: string,
    values?: unknown[]
  ): Promise<[T, unknown]>;
}

interface UserRow extends RowDataPacket {
  idUser: number;
  name: string;
  surname: string;
  email: string;
}

interface SessionRow extends RowDataPacket {
  idUser: number;
  name: string;
  surname: string;
  email: string;
  sid: string;
}

export async function log(
  db: DbExecutor,
  action: string,
  email?: string,
  userAgent?: string,
  ip?: string,
  accountId?: number
): Promise<number | null> {
  try {
    const sql = `
      INSERT INTO access_log (fk_account, ip, user_agent, email, operation)
      VALUES (?, ?, ?, ?, ?);
    `;
    const [result] = await db.execute<ResultSetHeader>(sql, [
      accountId ?? null,
      ip ?? null,
      userAgent ?? null,
      email ?? null,
      action
    ]);
    return result.insertId;
  } catch (error) {
    console.error("failed to write access log", error);
    return null;
  }
}

export async function insertUser(
  db: DbExecutor,
  name: string,
  surname: string,
  email: string,
  passwordHash: string
): Promise<number | null> {
  const sql = `
    INSERT INTO User(name, surname, email, password_hash)
    VALUES (?, ?, ?, ?);
  `;

  try {
    const [result] = await db.execute<ResultSetHeader>(sql, [name, surname, email, passwordHash]);
    await log(db, "user_registered", email, undefined, undefined, result.insertId);
    return result.insertId;
  } catch (error) {
    console.error("failed to insert user", error);
    return null;
  }
}

export async function insertSession(
  db: DbExecutor,
  userId: number,
  email?: string
): Promise<string | null> {
  const sessionId = uuidv4();
  const sql = "INSERT INTO session (sid, fk_user) VALUES (?, ?);";

  try {
    await db.execute<ResultSetHeader>(sql, [sessionId, userId]);
    await log(db, "session_created", email, undefined, undefined, userId);
    return sessionId;
  } catch (error) {
    console.error("failed to create session", error);
    return null;
  }
}

export async function accountLogin(
  db: DbExecutor,
  email: string,
  passwordHash: string
): Promise<AccountResponse | null> {
  const sql = `
    SELECT idUser, name, surname, email
    FROM User
    WHERE email = ? AND password_hash = ?;
  `;

  const [rows] = await db.execute<UserRow[]>(sql, [email, passwordHash]);
  const user = rows[0];

  if (!user) {
    await log(db, "failed_login", email);
    return null;
  }

  const sessionId = await insertSession(db, user.idUser, user.email);
  await log(db, "login_success", user.email, undefined, undefined, user.idUser);

  return {
    name: user.name,
    surname: user.surname,
    email: user.email,
    sessionID: sessionId
  };
}

export async function fetchSessionData(
  db: DbExecutor,
  sessionId: string
): Promise<AccountResponse | null> {
  const sql = `
    SELECT u.idUser, u.name, u.surname, u.email, s.sid
    FROM session s
    JOIN User u ON u.idUser = s.fk_user
    WHERE s.sid = ? AND (s.ttl IS NULL OR s.ttl > CURRENT_TIMESTAMP)
    LIMIT 1;
  `;

  const [rows] = await db.execute<SessionRow[]>(sql, [sessionId]);
  const row = rows[0];
  if (!row) {
    return null;
  }

  return {
    name: row.name,
    surname: row.surname,
    email: row.email,
    sessionID: row.sid
  };
}

File: auth_service/dbq.py
from uuid import uuid4


def log(db, action, email=None, user_agent=None, ip=None, account_id=None):
    cursor = db.cursor()
    sql = """
        INSERT INTO access_log (fk_account, ip, user_agent, email, operation)
        VALUES (%s, %s, %s, %s, %s);
    """
    cursor.execute(sql, (account_id, ip, user_agent, email, action))
    db.commit()
    return cursor.lastrowid


def insert_user(db, name, surname, email, password_hash):
    sql = """
        INSERT INTO User(name, surname, email, password_hash)
        VALUES (%s, %s, %s, %s);
    """
    cursor = db.cursor()
    try:
        cursor.execute(sql, (name, surname, email, password_hash))
        db.commit()
    except Exception:
        db.rollback()
        return None

    user_id = cursor.lastrowid
    if not user_id:
        return None

    try:
        log(db, "user_registered", email=email, account_id=user_id)
    except Exception:
        print("failed to log 'user_registered'")

    return user_id


def insert_session(db, uid, email=None):
    cursor = db.cursor()
    session_id = str(uuid4())
    try:
        sql = "INSERT INTO session (sid, fk_user) VALUES (%s, %s);"
        cursor.execute(sql, (session_id, uid))
        db.commit()
    except Exception:
        db.rollback()
        return None

    try:
        log(db, "session_created", email=email, account_id=uid)
    except Exception:
        print("failed to log 'session_created'")

    return session_id


def account_login(db, email, password_hash):
    cursor = db.cursor()
    sql = """
        SELECT idUser, name, surname, email
        FROM User
        WHERE email = %s AND password_hash = %s;
    """
    cursor.execute(sql, (email, password_hash))
    user = cursor.fetchone()

    if not user:
        try:
            log(db, "failed_login", email=email)
        except Exception:
            print("failed to log 'failed_login'")
        return None

    session_id = insert_session(db, user["idUser"], email=user["email"])
    account = {
        "name": user["name"],
        "surname": user["surname"],
        "email": user["email"],
    }
    if session_id:
        account["sessionID"] = session_id

    try:
        log(
            db,
            "login_success",
            email=user["email"],
            account_id=user["idUser"],
        )
    except Exception:
        print("failed to log 'login_success'")

    return account


def fetch_session_data(db, session_id):
    cursor = db.cursor()
    sql = """
        SELECT u.idUser, u.name, u.surname, u.email, s.sid
        FROM session s
        JOIN User u ON u.idUser = s.fk_user
        WHERE s.sid = %s AND (s.ttl IS NULL OR s.ttl > CURRENT_TIMESTAMP)
        LIMIT 1;
    """
    cursor.execute(sql, (session_id,))
    row = cursor.fetchone()

    if not row:
        return None

    return {
        "name": row["name"],
        "surname": row["surname"],
        "email": row["email"],
        "sessionID": row["sid"],
    }


